test/test.js

```diff
@@ -346,6 +346,104 @@ exports['constructor from promise - errors'] = function (test) {
         });
 };
 
+function createTestIterator(array, error) {
+    var count = 0,
+        length = array.length;
+    return {
+        next: function() {
+            if(count < length) {
+                if(error && count === 2) {
+                    throw error;
+                }
+                var iterElem = {
+                    value: array[count], done: false
+                };
+                count++;
+                return iterElem;
+            }
+            else {
+                return {
+                    value: void 0, done: true
+                }
+            }
+        }
+    };
+}
+
+exports['constructor from iterator'] = function (test) {
+    test.expect(1);
+    _(createTestIterator([1, 2, 3, 4, 5])).toArray(function (xs) {
+        test.same(xs, [1, 2, 3, 4, 5]);
+    });
+    test.done();
+};
+
+exports['constructor from iterator - error'] = function (test) {
+    test.expect(2);
+    _(createTestIterator([1, 2, 3, 4, 5], new Error('Error at index 2'))).errors(function (err) {
+        test.equals(err.message, 'Error at index 2');
+    }).toArray(function (xs) {
+        test.same(xs, [1, 2]);
+    });
+    test.done();
+};
+
+//ES6 iterators Begin
+if(global.Map && global.Symbol) {
```

Comment:
Actually, that's a good point we should just default to the Symbol.iterator property if we're given an iterable. I think it makes sense to support both as they're two sides of the same coin. One thing though, when I do something like:

```javascript
else if (typeof Symbol !== 'undefined' && xs[Symbol.iterator]) {
            return iteratorGenerator(xs[Symbol.iterator]);
        }
```

ESLint gives me an error saying the Symbol is not defined, should I be checking for global variables using the ```_.global``` object?

Also, I think you're right about the deoptimisation only applying to the inner function, I was misremembering that Bluebird performance doc.